From 083a6582f5ce8745735a4d90bd3daa8b29f9d31f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Carl=20=C3=85stholm?= <carl@astholm.se>
Date: Wed, 5 Nov 2025 16:34:07 +0100
Subject: [PATCH 048/133] link.Elf: Ensure archive header fields are not left
 blank
X-Developer-Signature: v=1; a=openpgp-sha256; l=1198; i=xtex@astrafall.org;
 h=from:subject; bh=kaePBSFk9PJtGXLyVuVsCJwlQU8lTaANgVzOsvkwAM4=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn78hiSZXe5el3cuZfrhs6h3skb5kwQ7Oz/7LIj7cPCy
 W43217d6ZjIwiDGxWAppshSZNjgzaqTzi+6rFwWZg4rE8gQaZEGBgYGBhYGvtzEvFIjHSM9U21D
 PUNDHSMdIwYuTgGY6ippRoZtGyzeHstYqS1cH7OsvP/Bg9vLvA+avP/sM9MyznK/+Olwhv9+u8I
 D357JnKWQdmHexzP+PDPMjY7JBdvtt/Pa0Jshmc4AAA==
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

ld.lld fails with "truncated or malformed archive" errors when reading
archive header field values that are not valid numbers.

(cherry picked from commit 43eb9b52ccac894023800115e8395cdb413be14a)
---
 src/link/Elf/Archive.zig | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/src/link/Elf/Archive.zig b/src/link/Elf/Archive.zig
index 2243dc478109..526caa094801 100644
--- a/src/link/Elf/Archive.zig
+++ b/src/link/Elf/Archive.zig
@@ -120,7 +120,6 @@ pub fn setArHdr(opts: struct {
         .ar_fmag = undefined,
     };
     @memset(mem.asBytes(&hdr), 0x20);
-    @memcpy(&hdr.ar_fmag, elf.ARFMAG);
 
     {
         var stream = std.io.fixedBufferStream(&hdr.ar_name);
@@ -132,10 +131,15 @@ pub fn setArHdr(opts: struct {
             .name_off => |x| writer.print("/{d}", .{x}) catch unreachable,
         }
     }
+    hdr.ar_date[0] = '0';
+    hdr.ar_uid[0] = '0';
+    hdr.ar_gid[0] = '0';
+    hdr.ar_mode[0] = '0';
     {
         var stream = std.io.fixedBufferStream(&hdr.ar_size);
         stream.writer().print("{d}", .{opts.size}) catch unreachable;
     }
+    hdr.ar_fmag = elf.ARFMAG.*;
 
     return hdr;
 }
-- 
2.52.0

